File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    public String[] solution(String[][] tickets) {
        Map<String, PriorityQueue<String>> fromToMap = new HashMap();
        
        for(String[] ticket : tickets){
            fromToMap.putIfAbsent(ticket[0], new PriorityQueue<>());
            fromToMap.get(ticket[0]).add(ticket[1]);
        }
        
        List<String> result = new LinkedList<>();
        Deque<String> stack = new ArrayDeque<>();
        
        stack.push("ICN");
        while(!stack.isEmpty()){
            while(fromToMap.containsKey(stack.peek()) && 
                  !fromToMap.get(stack.peek()).isEmpty()){
                stack.push(fromToMap.get(stack.peek()).poll());
            }
            
            result.add(0,stack.pop());
        }
        
        return result.toArray(new String[0]);
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.*

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val fromToMap: MutableMap<String, PriorityQueue<String>> = mutableMapOf()
        
        for(ticket in tickets){
            fromToMap.putIfAbsent(ticket[0], PriorityQueue())
            fromToMap[ticket[0]]!!.add(ticket[1])
        }
        
        val stack: Deque<String> = ArrayDeque()
        var answer:LinkedList<String> = LinkedList()
        
        stack.push("ICN")
        while(!stack.isEmpty()){
            while(fromToMap.containsKey(stack.peek()) && 
                    !fromToMap[stack.peek()]!!.isEmpty()){
                stack.push(fromToMap[stack.peek()]!!.poll())
            }
            answer.add(0,stack.pop())
        }
        
        return answer.toTypedArray()
    }
}